src/main/kotlin/edu/kit/compiler/semantic/visitor/NameAnalysis.kt

```diff
@@ -0,0 +1,481 @@
+            )
+            return null
+        }
+
+        val def = clazz.namespace.fields.getOrNull(name.symbol)
+        if (def == null) {
+            sourceFile.annotate(
+                AnnotationType.ERROR,
+                name.sourceRange,
+                "unknown variable `${name.symbol.text}`"
+            )
+            return null
+        }
+        return def.wrap()
+    }
+
+    fun lookupThis(): ClassDefinition = thisDefinition
+
+    fun hasAnyDefinitionForSystem(): Boolean {
+        return global.classes.getOrNull("System") != null ||
+            clazz.namespace.fields.getOrNull("System") != null ||
+            local.contains(systemSymbol)
+    }
+}
+
+/**
+ * First name analysis pass: Populate global namespace with all classes and populate class namespaces with contained fields and methods.
+ */
+@Suppress("SpellCheckingInspection")
+class NamespacePopulator(
+    private val global: GlobalNamespace,
+    private val sourceFile: AnnotatableFile,
+) : AbstractVisitor() {
+
+    private lateinit var currentClassNamespace: ClassNamespace
+
+    override fun visitClassDeclaration(classDeclaration: AstNode.ClassDeclaration) {
+        if (classDeclaration.name.text == "String") {
+            sourceFile.annotate(
+                AnnotationType.ERROR,
+                classDeclaration.name.sourceRange,
+                "cannot shadow built-in `String` class"
+            )
+            return
+        }
+
+        val namespace = ClassNamespace(global)
+
+        global.classes.tryPut(classDeclaration.asDefinition(), onDuplicate = {
+            sourceFile.annotate(
+                AnnotationType.ERROR,
+                classDeclaration.name.sourceRange,
+                "class `${classDeclaration.name.text}` is already defined",
+                listOf(
+                    SourceNote(it.node.name.sourceRange, "see previous definition here")
+                )
+            )
+        })
+
+        classDeclaration.namespace = namespace
+        currentClassNamespace = namespace
+
+        super.visitClassDeclaration(classDeclaration) // descend
+    }
+
+    override fun visitFieldDeclaration(fieldDeclaration: AstNode.ClassMember.FieldDeclaration) {
+        currentClassNamespace.fields.tryPut(fieldDeclaration.asDefinition(), onDuplicate = { prev ->
+            sourceFile.annotate(
+                AnnotationType.ERROR,
+                fieldDeclaration.name.sourceRange,
+                "field `${fieldDeclaration.name.text}` is already defined",
+                listOf(
+                    SourceNote(prev.node.name.sourceRange, "see previous definition here")
+                )
+            )
+        })
+
+        // do not descend
+    }
+
+    override fun visitMethodDeclaration(methodDeclaration: AstNode.ClassMember.SubroutineDeclaration.MethodDeclaration) {
+        if (methodDeclaration.name.text == "main") {
+            currentClassNamespace.mainMethodDefinition?.let {
```

Comment:
`let` ist eigentlich ein `map` call (Es gibt den Wert der letzten Expression zurück). Wenn man einfach nur nen Closure haben will, in dem der Receiver der Parameter wird, benutz `also`, das ist konsequenter und damit verständlicher (als returnwert gibt es immer den Receiver wieder zurück). Wenn man eines haben möchte, in dem der Receiver ein Receiver bleibt, benutz `apply` (es gibt ebenfalls den Receiver wieder zurück).

Mir nicht so wichtig, dass es hier geändert wird, der Code ist ja äquivalent, ich finds nur schöner für die Zukunft.